### DISCUSSION TITLE
Accessing the photoprism API via an external script

### DISCUSSION THREAD

**Discussion by doebi** [photoprism/photoprism#3692] [General]
Hi, so as the title suggests I want to access my photos stored in photoprism from an external script. The basic idea is to have a bash script retrieve a random picture and display it on a display. It is just for personal use within a secured private network, so no need for big security measures.

I found this guide https://docs.photoprism.app/developer-guide/api/ and it suggests to extract the `session_id` from your browser session to use it within a curl script. This works fine for me when calling `/api/v1/photos`, `/api/v1/photos/:uid`, etc.

But if I call `/api/v1/photos/:uid/dl` or `/api/v1/dl` I only get some svg returned.

In the code I traced it back to this piece
https://github.com/photoprism/photoprism/blob/afe219079711f14fba7ee47aee9e65afddaa6217/internal/api/photos.go#L130

It looks like as if my extracted token is note elligible to download.

Since the manual extraction of browser sessions is not a feasible solution in the long run, I am asking on how to authenticate to the API for my script. Any advice? Thanks.

**Comment by doebi** [comment]
Ok, I figured it out...

I came up with a small script, that performs login, searching, downloading and viewing of images. Have fun.

```bash
#!/usr/bin/env sh

base_url="http://127.0.0.1:2342"

# Credentials
username=""
password=""

# Creating Session
response=$(curl -s -H "Content-Type: application/json" -d '{"username": "'$username'", "password": "'$password'"}' $base_url"/api/v1/session")
session_id=$(echo $response | jq -r ".id")

# Alternatively supply a $session_id directly
# session_id=

# Find Images
images=$(curl -s $base_url"/api/v1/photos/view?count=720&offset=0&merged=true&country=&camera=2&lens=0&label=&year=0&month=0&color=&order=newest&q=&public=true&quality=3" \
  -H "X-Session-ID: "$session_id)

# Select first image
selected=$(echo $images | jq ".[0]")

# Retrieve download url
download_url=$(echo $selected | jq -r ".DownloadUrl")

# download image
curl -s -o temp.jpg $base_url$download_url

eog temp.jpg &
                     

```

**Reply by kuleshov-aleksei** [reply]
Thank you for your script!

I found it while searching for references how to use photoprism's API and want to leave a trail for someone else

Slightly modified it for running auto import via cron (since photoprism does supports auto import **only** when using WebDAV)

```bash
#!/bin/bash

set -e

base_url="http://127.0.0.1:2342"

# Credentials
username=""
password=""

# Creating Session
response=$(curl -s -H "Content-Type: application/json" -d '{"username": "'$username'", "password": "'$password'"}' $base_url"/api/v1/session")
session_id=$(echo $response | jq -r ".id")

# Run import
output=$(curl -s --request POST \
  --url $base_url"/api/v1/import/" \
  --header 'Content-Type: application/json' \
  --header "X-Session-ID: "$session_id \
  --data '{
    "path": "",
    "move": true
}')

echo $output | jq -r ".message"
```

data in import request can be modified for importing from specific location or not moving files away from import location

Run every 5 minutes:
```
*/5 * * * * /root/crons/auto_import.sh
```

**Reply by lastzero** [reply]
Note that our latest release supports app passwords that you can use as access tokens, see our updated developer guide: https://docs.photoprism.app/developer-guide/api/#client-authentication

**Reply by alittler** [reply]
And **I** edited it to reproduce the Session ID into Terminal, because I needed it for my home server integration.

`#!/bin/bash

set -e

base_url="http://127.0.0.1:2342"

# Credentials		
username=""
password=""

# Creating Session
response=$(curl -s -H "Content-Type: application/json" -d '{"username": "'$username'", "password": "'$password'"}' $base_url"/api/v1/session")
session_id=$(echo $response | jq -r ".id")

# Output the session_id
echo "Session ID: $session_id"

# Run import
output=$(curl -s --request POST \
  --url $base_url"/api/v1/import/" \
  --header 'Content-Type: application/json' \
  --header "X-Session-ID: "$session_id \
  --data '{
    "path": "",
    "move": true
}')

echo $output | jq -r ".message"
`

**Reply by lastzero** [reply]
If you create an app password or access token using the CLI commands that PhotoPrism provides, you won't need to request a new token through the create session endpoint (and you can simply submit it with a standard Authorization header) as described in our developer guide:
- https://docs.photoprism.app/developer-guide/api/#client-authentication

**Comment by doebi** [comment]
One note though: At the moment the script assumes, that the selected image is a type 'image/jpeg'. To make this more robust, it should retrieve the file metadata first and extract a filetype or even filename from there. But this requires an additional call since file metadata comes along the `/api/v1/photos` endpoint and download urls come via `/api/v1/photos/view` ...

**Comment by THF-Mediapics** [comment]
Are the API endpoints generally available to all users? I only have Photoprism Plus and would like to work with API. is this possible?

**Reply by lastzero** [reply]
Yes 👍

**Reply by THF-Mediapics** [reply]
that's great but I keep getting an error 400 back {“code”:400, “error”: “This is not possible”} even though I provide an access token.

**Reply by lastzero** [reply]
Have you checked our API documentation and Developer Guide for hints and examples yet?

- https://docs.photoprism.dev/
- https://docs.photoprism.app/developer-guide/api/#client-authentication
